source/agora/consensus/EnrollmentPool.d

```diff
@@ -0,0 +1,404 @@
+
+        Params:
+            enroll = the enrollment data to add
+            finder = the delegate to find UTXOs with
+
+        Returns:
+            true if the enrollment data has been added to the enrollment pool
+
+    ***************************************************************************/
+
+    public bool add (const ref Enrollment enroll, scope UTXOFinder finder)
+        @safe nothrow
+    {
+        // check validity of the enrollment data
+        if (auto reason = isInvalidEnrollmentReason(enroll, finder))
+        {
+            log.info("Invalid enrollment data: {}, Data was: {}", reason, enroll);
+            return false;
+        }
+
+        // check if already exists
+        if (this.hasEnrollment(enroll.utxo_key))
+        {
+            log.info("Rejected already existing enrollment, Data was: {}",
+                enroll);
+            return false;
+        }
+
+        static ubyte[] buffer;
+        try
+        {
+            serializeToBuffer(enroll, buffer);
+        }
+        catch (Exception ex)
+        {
+            log.error("Serialization error: {}, Data was: {}", ex.msg, enroll);
+            return false;
+        }
+
+        try
+        {
+            () @trusted {
+                this.db.execute("INSERT INTO enrollment_pool (key, val) VALUES (?, ?)",
+                    enroll.utxo_key.toString(), buffer);
+            }();
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error: {}, Data was: {}", ex.msg, enroll);
+            return false;
+        }
+
+        return true;
+    }
+
+    /***************************************************************************
+
+        Returns:
+            the number of enrollments being managed by the enrollment pool,
+            which are un-registered enrollments.
+
+    ***************************************************************************/
+
+    public size_t count () @trusted nothrow
+    {
+        size_t cnt;
+        try
+        {
+            cnt = this.db.execute("SELECT count(*) FROM enrollment_pool").
+                oneValue!size_t;
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error on count");
+        }
+        return cnt;
+    }
+
+    /***************************************************************************
+
+        Remove the enrollment data with the given key from the enrollment pool
+
+        Params:
+            enroll_hash = key for an enrollment data to remove
+
+    ***************************************************************************/
+
+    public void remove (const ref Hash enroll_hash) @trusted nothrow
+    {
+        try
+        {
+            this.db.execute("DELETE FROM enrollment_pool WHERE key = ?",
+                enroll_hash.toString());
+        }
+        catch (Exception ex)
+        {
+            log.error("Database operation error on remove");
```

Comment:
When I said before that the API needs to be consistent I meant the code should do either one of the following:

A) Return a status code for every function which can fail. So the calling code can handle it.
B) Make all functions throw if they can fail. The calling code then must use a try/catch to handle it, or it can let the exception propagate upwards so other calling code handles it.
C) Catch any exceptions and halt the program with `assert(0)`.

Right now the code doesn't do any of these. Instead it catches the exception, logs it, and continue execution.

But what is the purpose of continuing execution like nothing bad happened? Will the database automatically fix itself on the next call? What is the expectation here?

Reply:
Removing the enrollment from the pool is not critical and so execution can continue. In this PR, if adding the enrollment to the validator set fails, the ledger makes the system terminate as follows.
`
foreach (enrollment; block.header.enrollments)
        if (!this.enroll_man.addValidator(enrollment,
                block.header.height, this.utxo_set.getUTXOFinder()))
            assert(0);
`
Mathias and I discussed that in the pair programming session.
The recovery from the database failure can be executed when a node start which is not implemented yet and can be another issue.